tests/unittests/Foundation/NSDecimalNumberTests.mm

```diff
@@ -0,0 +1,332 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <TestFramework.h>
+#import <Foundation/Foundation.h>
+#import <cmath>
+
+#define EXPECT_ARRAY_EQUAL(a, b)                     \
+    do {                                             \
+        for (int i = 0; i < NSDecimalMaxSize; ++i) { \
+            EXPECT_EQ(a[i], b[i]);                   \
+        }                                            \
+    } while (0, 0)
+
+static void testSubtract(NSDecimalNumber* x, NSDecimalNumber* y, NSDecimalNumber* expected) {
+    NSDecimalNumber* result = [x decimalNumberBySubtracting:y];
+    EXPECT_OBJCEQ(expected, result);
+}
+
+static void testAddition(NSDecimalNumber* x, NSDecimalNumber* y, NSDecimalNumber* expected) {
+    NSDecimalNumber* result = [x decimalNumberByAdding:y];
+    EXPECT_OBJCEQ(expected, result);
+}
+
+#define VALIDATE(decimal, len, exp, isNegative, mantissa) \
+    EXPECT_EQ(len, decimal._length);                      \
+    EXPECT_EQ(exp, decimal._exponent);                    \
+    EXPECT_EQ(isNegative, decimal._isNegative);           \
+    EXPECT_ARRAY_EQUAL(mantissa, decimal._mantissa);
+
+TEST(NSDecimalNumber, Limits) {
+    // SCHAR_MIN
+    unsigned short mantissaSCHAR_MIN[] = { 128, 0, 0, 0, 0, 0, 0, 0, 0 };
+    NSDecimalNumber* dNumber = [[[NSDecimalNumber alloc] initWithShort:SCHAR_MIN] autorelease];
+    NSDecimal decimal = [dNumber decimalValue];
+    VALIDATE(decimal, 1, 0, 1, mantissaSCHAR_MIN);
+
+    // SCHAR_MAX
+    unsigned short mantissaSCHAR_MAX[] = { 127, 0, 0, 0, 0, 0, 0, 0, 0 };
+    dNumber = [[[NSDecimalNumber alloc] initWithShort:SCHAR_MAX] autorelease];
+    decimal = [dNumber decimalValue];
+    VALIDATE(decimal, 1, 0, 0, mantissaSCHAR_MAX);
+
+    // LLONG_MIN
+    unsigned short mantissaLLMin[] = { 0, 0, 0, 32768, 0, 0, 0, 0, 0 };
+    dNumber = [[[NSDecimalNumber alloc] initWithLongLong:LLONG_MIN] autorelease];
+    decimal = [dNumber decimalValue];
+    VALIDATE(decimal, 4, 0, 1, mantissaLLMin);
+
+    // LLONG_MAX
+    unsigned short mantissaLLMax[] = { 65535, 65535, 65535, 32767, 0, 0, 0, 0, 0 };
+    dNumber = [[[NSDecimalNumber alloc] initWithLongLong:LLONG_MAX] autorelease];
+    decimal = [dNumber decimalValue];
+    VALIDATE(decimal, 4, 0, 0, mantissaLLMax);
```

Comment:
not always, often they might differ. e.g SHRT example.
secondly, MIN becomes 0x8000000000000000 just before it's set as the mantissa, while max becomes 0x7fffffffffffffff 

---
In reply to: [80737862](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r80737862) [](ancestors = 80737862)